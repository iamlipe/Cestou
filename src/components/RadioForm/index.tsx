import React, {memo, useState} from 'react';
import styled, {useTheme} from 'styled-components/native';
import Icon from 'react-native-vector-icons/MaterialIcons';
import {useController} from 'react-hook-form';

interface Props {
  name: string;
  control: any;
  options: string[];
}

const CheckBox = ({name, control, options}: Props) => {
  const [checkeds, setCheckeds] = useState<string>('');
  const theme = useTheme();

  const {
    field: {onChange},
  } = useController({name, control});

  const renderOptions = (option: string, index: number) => (
    <StyledContainerRadio
      key={index}
      onPress={() => {
        setCheckeds(option);
        onChange(option);
      }}>
      <Icon
        name={
          checkeds.includes(option)
            ? 'radio-button-checked'
            : 'radio-button-unchecked'
        }
        testID={
          checkeds.includes(option)
            ? `radio-icon-checked-${index}`
            : `radio-icon-not-checked-${index}`
        }
        color={
          checkeds.includes(option)
            ? theme.colors.PRIMARY_800
            : theme.colors.GRAY_700
        }
        size={24}
      />
      <StyledTitle>{option}</StyledTitle>
    </StyledContainerRadio>
  );

  return (
    <>
      {options.map((option, index) => {
        return renderOptions(option, index);
      })}
    </>
  );
};

const StyledContainerRadio = styled.TouchableOpacity`
  flex-direction: row;
  align-items: center;
  margin-bottom: 10px;
`;

const StyledTitle = styled.Text`
  font-family: ${({theme}) => theme.fonts.REGULAR_SOURCESANSPRO};
  font-size: ${({theme}) => theme.sizing.SMALLER};
  color: ${({theme}) => theme.colors.GRAY_900};
  margin-left: 5px;
`;

export default memo(CheckBox);
