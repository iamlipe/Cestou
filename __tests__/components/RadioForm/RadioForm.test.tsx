import React from 'react';
import {
  renderWithThemeProvider,
  matchSnapshotWithProvider,
} from '@/config/functions/testing';

import RadioForm from '@/components/RadioForm';
import {fireEvent} from '@testing-library/react-native';

jest.mock('react-hook-form', () => ({
  useController: () => ({
    field: {
      onChange: jest.fn(),
    },
  }),
}));

describe('RadioForm', () => {
  const options = ['radio-me', 'radio-me-too'];

  test('should render correctly', () => {
    renderWithThemeProvider(
      <RadioForm name="radio" control={jest.fn()} options={options} />,
    );
  });

  test('should matches snapshot', () => {
    matchSnapshotWithProvider(
      <RadioForm name="radio" control={jest.fn()} options={options} />,
    );
  });

  test('should render all options', () => {
    const radio = renderWithThemeProvider(
      <RadioForm name="radio" control={jest.fn()} options={options} />,
    );

    options.forEach(option => {
      expect(radio.getByText(option)).toBeTruthy();
    });
  });

  test('should starting with no option selected', () => {
    const radio = renderWithThemeProvider(
      <RadioForm name="radio" control={jest.fn()} options={options} />,
    );

    options.forEach((_option, index) => {
      const notCheckedIcon = radio.queryByTestId(
        `radio-icon-not-checked-${index}`,
      );

      expect(notCheckedIcon).toBeTruthy();
    });
  });

  test('should select option', () => {
    const radio = renderWithThemeProvider(
      <RadioForm name="radio" control={jest.fn()} options={options} />,
    );

    options.forEach((_option, index) => {
      const notCheckedIcon = radio.getByTestId(
        `radio-icon-not-checked-${index}`,
      );

      expect(notCheckedIcon).toBeTruthy();

      fireEvent.press(notCheckedIcon);

      const radioIcon = radio.queryByTestId(`radio-icon-checked-${index}`);

      expect(radioIcon).toBeTruthy();
    });
  });
});
