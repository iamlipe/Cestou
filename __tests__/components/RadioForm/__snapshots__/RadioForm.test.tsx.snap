// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioForm should matches snapshot 1`] = `
Array [
  <View
    style={
      Array [
        Object {},
      ]
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginBottom": 10,
          "opacity": 1,
        }
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          Array [
            Object {
              "color": "#3A3A3A",
              "fontSize": 24,
            },
            undefined,
            Object {
              "fontFamily": "Material Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
        testID="radio-icon-not-checked-0"
      >
        
      </Text>
      <View
        style={
          Array [
            Object {
              "alignItems": "flex-start",
              "flexDirection": "column",
            },
          ]
        }
      >
        <Text
          detailsOptions={false}
          style={
            Array [
              Object {
                "color": "#1A1A1A",
                "fontFamily": "SourceSansPro-Regular",
                "fontSize": 27,
                "marginLeft": 8,
              },
            ]
          }
        >
          radio-me
        </Text>
      </View>
    </View>
  </View>,
  <View
    style={
      Array [
        Object {},
      ]
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginBottom": 10,
          "opacity": 1,
        }
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          Array [
            Object {
              "color": "#3A3A3A",
              "fontSize": 24,
            },
            undefined,
            Object {
              "fontFamily": "Material Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
        testID="radio-icon-not-checked-1"
      >
        
      </Text>
      <View
        style={
          Array [
            Object {
              "alignItems": "flex-start",
              "flexDirection": "column",
            },
          ]
        }
      >
        <Text
          detailsOptions={false}
          style={
            Array [
              Object {
                "color": "#1A1A1A",
                "fontFamily": "SourceSansPro-Regular",
                "fontSize": 27,
                "marginLeft": 8,
              },
            ]
          }
        >
          radio-me-too
        </Text>
      </View>
    </View>
  </View>,
  <Text
    style={
      Array [
        Object {
          "alignSelf": "flex-end",
          "color": "#781B1C",
          "fontFamily": "SourceSansPro-Regular",
          "fontSize": 24,
          "marginBottom": 16,
          "marginLeft": 0,
          "marginRight": 0,
          "marginTop": -8,
        },
      ]
    }
  />,
]
`;
